UM/Qt/QtRenderer.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2015 Ultimaker B.V.
+# Uranium is released under the terms of the AGPLv3 or higher.
+
+from PyQt5.QtGui import QColor, QOpenGLBuffer, QOpenGLContext, QOpenGLFramebufferObject, QOpenGLFramebufferObjectFormat, QSurfaceFormat, QOpenGLVersionProfile, QImage
+
+from UM.Application import Application
+from UM.View.Renderer import Renderer
+from UM.Math.Vector import Vector
+from UM.Math.Matrix import Matrix
+from UM.Resources import Resources
+from UM.Logger import Logger
+from UM.Scene.Iterator.DepthFirstIterator import DepthFirstIterator
+from UM.Scene.Selection import Selection
+from UM.Scene.PointCloudNode import PointCloudNode
+from UM.Math.Color import Color
+
+from UM.Mesh.MeshBuilder import MeshBuilder
+from UM.View.CompositePass import CompositePass
+from UM.View.DefaultPass import DefaultPass
+from UM.View.SelectionPass import SelectionPass
+from UM.View.GL.OpenGL import OpenGL
+from UM.View.RenderBatch import RenderBatch
+from UM.Qt.GL.QtOpenGL import QtOpenGL
+
+import numpy
+import copy
+from ctypes import c_void_p
+
+vertexBufferProperty = "__qtgl2_vertex_buffer"
+indexBufferProperty = "__qtgl2_index_buffer"
+
+##  A Renderer implementation using PyQt's OpenGL implementation to render.
+class QtRenderer(Renderer):
+    def __init__(self):
+        super().__init__()
+
+        self._controller = Application.getInstance().getController()
+        self._scene = self._controller.getScene()
+
+        self._vertex_buffer_cache = {}
+        self._index_buffer_cache = {}
+
+        self._initialized = False
+
+        self._light_position = Vector(0, 0, 0)
+        self._background_color = QColor(128, 128, 128)
+        self._viewport_width = 0
+        self._viewport_height = 0
+        self._window_width = 0
+        self._window_height = 0
+
+        self._batches = []
+
+        self._quad_buffer = None
+
+        self._camera = None
+
+    ##  Get an integer multiplier that can be used to correct for screen DPI.
+    def getPixelMultiplier(self):
+        # Standard assumption for screen pixel density is 96 DPI. We use that as baseline to get
+        # a multiplication factor we can use for screens > 96 DPI.
+        return round(Application.getInstance().primaryScreen().physicalDotsPerInch() / 96.0)
+
+    ##  Get the list of render batches.
+    def getBatches(self):
+        return self._batches
+
+    ##  Overridden from Renderer.
+    #
+    #   This makes sure the added render pass has the right size.
+    def addRenderPass(self, render_pass):
+        super().addRenderPass(render_pass)
+        render_pass.setSize(self._viewport_width, self._viewport_height)
+
+    ##  Set background color used when rendering.
```

Comment:
Renderer currently has no viewport size, thus cannot use those values. Arguably, viewport/window size could be moved to renderer as well.